File: frontend/src/api/redemptions.ts
/**
 * The redemption API client, for the merged `/api/v1/redemptions/` routes
 * (issue #55): a child spends points on a reward, and a parent fulfils or
 * cancels the request. Shaped like `api/rewards.ts` and `api/chores.ts`:
 * same-origin session authentication and CSRF only, and every response
 * validated at runtime before it is trusted.
 *
 * The child shape (`ChildRedemption`) and the parent shape
 * (`ParentRedemption`, which adds who asked) are two different response
 * bodies from the one `GET` route, exactly as the server's two serializers
 * differ; this client validates whichever one the caller asks for.
 */

import { CSRF_HEADER_NAME, ensureCsrfToken } from './session'

export type RedemptionStatus = 'pending' | 'fulfilled' | 'cancelled'

export interface ChildRedemption {
  id: number
  reward_name: string
  reward_points: number
  status: RedemptionStatus
  created_at: string
  decided_at: string | null
}

export interface ParentRedemption extends ChildRedemption {
  child_id: number
  child_username: string
}

export const REDEMPTIONS_QUERY_KEY = ['redemptions'] as const

/**
 * - `validation`: the request body was refused, most often because the
 *   reward is unaffordable. `fieldErrors` carries the first message per
 *   field the server named.
 * - `forbidden`: the caller may not perform this action in this household.
 * - `notFound`: the reward or redemption does not exist, belongs to another
 *   household, or (for fulfil and cancel) was already decided.
 * - `unavailable`: unreachable, timed out, not JSON, or an unexpected status.
 */
export type RedemptionFailureKind =
  | 'validation'
  | 'forbidden'
  | 'notFound'
  | 'unavailable'

export class RedemptionRequestError extends Error {
  readonly kind: RedemptionFailureKind
  readonly fieldErrors: Readonly<Record<string, string>>

  constructor(kind: RedemptionFailureKind, fieldErrors: Record<string, string> = {}) {
    super(`Redemption request failed: ${kind}`)
    this.name = 'RedemptionRequestError'
    this.kind = kind
    this.fieldErrors = fieldErrors
  }
}

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === 'object' && value !== null && !Array.isArray(value)

const isJsonMediaType = (contentType: string | null) =>
  contentType !== null &&
  /^application\/(?:[a-z0-9!#$&^_.+-]+\+)?json(?:\s*;|$)/i.test(contentType)

const isRedemptionStatus = (value: unknown): value is RedemptionStatus =>
  value === 'pending' || value === 'fulfilled' || value === 'cancelled'

const hasChildShape = (value: Record<string, unknown>): boolean =>
  typeof value.id === 'number' &&
  Number.isInteger(value.id) &&
  typeof value.reward_name === 'string' &&
  typeof value.reward_points === 'number' &&
  Number.isInteger(value.reward_points) &&
  isRedemptionStatus(value.status) &&
  typeof value.created_at === 'string' &&
  (value.decided_at === null || typeof value.decided_at === 'string')

export const isChildRedemption = (value: unknown): value is ChildRedemption =>
  isRecord(value) && Object.keys(value).length === 6 && hasChildShape(value)

export const isParentRedemption = (value: unknown): value is ParentRedemption =>
  isRecord(value) &&
  Object.keys(value).length === 8 &&
  hasChildShape(value) &&
  typeof value.child_id === 'number' &&
  Number.isInteger(value.child_id) &&
  typeof value.child_username === 'string'

const isChildRedemptionList = (value: unknown): value is ChildRedemption[] =>
  Array.isArray(value) && value.every(isChildRedemption)

const isParentRedemptionList = (value: unknown): value is ParentRedemption[] =>
  Array.isArray(value) && value.every(isParentRedemption)

const readFieldErrors = (body: unknown): Record<string, string> => {
  if (!isRecord(body)) {
    return {}
  }
  const fieldErrors: Record<string, string> = {}
  for (const [field, messages] of Object.entries(body)) {
    if (
      Array.isArray(messages) &&
      messages.length > 0 &&
      typeof messages[0] === 'string'
    ) {
      fieldErrors[field] = messages[0]
    }
  }
  return fieldErrors
}

const readJsonBody = async (response: Response): Promise<unknown> => {
  if (!isJsonMediaType(response.headers.get('Content-Type'))) {
    return undefined
  }
  try {
    return await response.json()
  } catch {
    return undefined
  }
}

const send = async (path: string, init: RequestInit): Promise<Response> => {
  try {
    return await fetch(path, { credentials: 'same-origin', ...init })
  } catch {
    throw new RedemptionRequestError('unavailable')
  }
}

/** `GET /api/v1/redemptions/`: the child's own history, newest first. */
export const fetchChildRedemptions = async ({
  signal,
}: {
  signal?: AbortSignal
}): Promise<ChildRedemption[]> => {
  const response = await send('/api/v1/redemptions/', {
    method: 'GET',
    headers: { Accept: 'application/json' },
    signal,
  })

  if (response.redirected) {
    throw new RedemptionRequestError('unavailable')
  }
  if (response.status === 403) {
    throw new RedemptionRequestError('forbidden')
  }
  const body = await readJsonBody(response)
  if (response.status !== 200 || !isChildRedemptionList(body)) {
    throw new RedemptionRequestError('unavailable')
  }
  return body
}

/** `GET /api/v1/redemptions/`: every household redemption, with who asked. */
export const fetchParentRedemptions = async ({
  signal,
}: {
  signal?: AbortSignal
}): Promise<ParentRedemption[]> => {
  const response = await send('/api/v1/redemptions/', {
    method: 'GET',
    headers: { Accept: 'application/json' },
    signal,
  })

  if (response.redirected) {
    throw new RedemptionRequestError('unavailable')
  }
  if (response.status === 403) {
    throw new RedemptionRequestError('forbidden')
  }
  const body = await readJsonBody(response)
  if (response.status !== 200 || !isParentRedemptionList(body)) {
    throw new RedemptionRequestError('unavailable')
  }
  return body
}

const readChildRedemptionBody = async (response: Response): Promise<ChildRedemption> => {
  if (response.redirected) {
    throw new RedemptionRequestError('unavailable')
  }
  const body = await readJsonBody(response)

  if (response.status === 400) {
    throw new RedemptionRequestError('validation', readFieldErrors(body))
  }
  if (response.status === 403) {
    throw new RedemptionRequestError('forbidden')
  }
  if (response.status === 404) {
    throw new RedemptionRequestError('notFound')
  }
  if ((response.status !== 200 && response.status !== 201) || !isChildRedemption(body)) {
    throw new RedemptionRequestError('unavailable')
  }
  return body
}

/**
 * `POST /api/v1/redemptions/`: spend points on `rewardId`.
 *
 * `idempotencyKey` must be a fresh value per redemption attempt, and the
 * same value again for a retry of that same attempt (a lost response, a
 * double-tapped button): the server replays the first result rather than
 * charging twice.
 */
export const redeemReward = async ({
  csrfToken,
  rewardId,
  idempotencyKey,
  signal,
}: {
  csrfToken: string
  rewardId: number
  idempotencyKey: string
  signal?: AbortSignal
}): Promise<ChildRedemption> =>
  readChildRedemptionBody(
    await send('/api/v1/redemptions/', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        [CSRF_HEADER_NAME]: csrfToken,
      },
      body: JSON.stringify({ reward: rewardId, idempotency_key: idempotencyKey }),
      signal,
    }),
  )

const readParentRedemptionBody = async (
  response: Response,
): Promise<ParentRedemption> => {
  if (response.redirected) {
    throw new RedemptionRequestError('unavailable')
  }
  const body = await readJsonBody(response)

  if (response.status === 403) {
    throw new RedemptionRequestError('forbidden')
  }
  if (response.status === 404) {
    throw new RedemptionRequestError('notFound')
  }
  if (response.status !== 200 || !isParentRedemption(body)) {
    throw new RedemptionRequestError('unavailable')
  }
  return body
}

const postAction = async (
  path: string,
  { csrfToken, signal }: { csrfToken: string; signal?: AbortSignal },
): Promise<ParentRedemption> =>
  readParentRedemptionBody(
    await send(path, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        [CSRF_HEADER_NAME]: csrfToken,
      },
      signal,
    }),
  )

/** `POST /api/v1/redemptions/<id>/fulfil/`: hand over the reward. Parent only. */
export const fulfilRedemption = async (
  args: { csrfToken: string; id: number; signal?: AbortSignal },
): Promise<ParentRedemption> => postAction(`/api/v1/redemptions/${args.id}/fulfil/`, args)

/** `POST /api/v1/redemptions/<id>/cancel/`: refuse and refund. Parent only. */
export const cancelRedemption = async (
  args: { csrfToken: string; id: number; signal?: AbortSignal },
): Promise<ParentRedemption> => postAction(`/api/v1/redemptions/${args.id}/cancel/`, args)

export { ensureCsrfToken }

File: chorum_murohc/api/redemptions.py
"""Redemption endpoints: a child spends points on a reward, and a parent
settles the request.

Three routes under `/api/v1/redemptions/`, same-origin session authenticated
and scoped to the one household the caller is a live member of:

- `GET redemptions/` lists redemptions. A child sees only their own, every
  status, newest first. A parent sees every redemption in the household, the
  same shape plus who asked, so the fulfilment queue and its history are one
  list.
- `POST redemptions/` redeems a reward: debits the points and creates the
  pending row in one transaction. Child only. A repeated `idempotency_key`
  replays the same row instead of charging twice.
- `POST redemptions/<pk>/fulfil/` marks a pending redemption fulfilled.
  Parent only.
- `POST redemptions/<pk>/cancel/` cancels a pending redemption and refunds
  its points in full, as a new ledger entry. Parent only.

All the actual state change, ledger and audit work happens in
`chorum_murohc.rewards.services`, re-validated inside its own transaction
immediately before writing. This module's job is authority, HTTP shape and
mapping the service's outcomes and exceptions onto responses; it holds no
transaction and no ledger write of its own.

A child can never fulfil or cancel: `IsHouseholdParent` denies the two action
routes before either view method runs, exactly as `chores.py` denies write
routes to a child. There is no reject and no reverse: fulfilled is final, and
refusing a pending request is cancelling it.
"""

from rest_framework import serializers, status
from rest_framework.authentication import SessionAuthentication
from rest_framework.exceptions import NotFound, PermissionDenied
from rest_framework.renderers import JSONRenderer
from rest_framework.response import Response
from rest_framework.views import APIView

from chorum_murohc.api.permissions import (
    PERMISSION_DENIED_DETAIL,
    IsHouseholdChild,
    IsHouseholdParent,
)
from chorum_murohc.api.session import resolve_active_membership
from chorum_murohc.identity.models import Membership
from chorum_murohc.rewards.models import Redemption, Reward
from chorum_murohc.rewards.services import (
    InsufficientPointsError,
    RewardAuthorityError,
    cancel_redemption,
    fulfil_redemption,
    redeem_reward,
)

# The one detail returned when a redemption would take the balance below
# zero. It names no balance and no reward.
INSUFFICIENT_POINTS_DETAIL = 'Not enough points for this reward.'

IDEMPOTENCY_KEY_MAX_LENGTH = 255


IsHouseholdParentOrChild = IsHouseholdParent | IsHouseholdChild


class RedemptionChildSerializer(serializers.ModelSerializer):
    """What a child sees for one of their own redemptions.

    `reward_name` and `reward_points` are the snapshot taken at redemption,
    so this reads correctly even after the catalogue reward is edited or
    removed. There is no live `reward` reference here: a child cannot browse
    to another household's reward through it, and there is nothing to browse
    to once the reward is gone.
    """

    class Meta:
        model = Redemption
        fields = (
            'id',
            'reward_name',
            'reward_points',
            'status',
            'created_at',
            'decided_at',
        )
        read_only_fields = fields


class RedemptionParentSerializer(serializers.ModelSerializer):
    """What a parent sees: the same shape plus who asked."""

    child_id = serializers.IntegerField(read_only=True)
    child_username = serializers.CharField(source='child.username', read_only=True)

    class Meta:
        model = Redemption
        fields = (
            'id',
            'child_id',
            'child_username',
            'reward_name',
            'reward_points',
            'status',
            'created_at',
            'decided_at',
        )
        read_only_fields = fields


class RedemptionWriteSerializer(serializers.Serializer):
    """The only writable body: which reward, and a client-chosen retry key.

    `reward` names the catalogue row by identifier; the service resolves and
    locks it within the caller's own household, so no queryset is built here
    and a foreign or unknown identifier surfaces as the service's own
    `Reward.DoesNotExist`, never a validation error that would confirm or
    deny its existence.
    """

    reward = serializers.IntegerField(min_value=1)
    idempotency_key = serializers.CharField(
        max_length=IDEMPOTENCY_KEY_MAX_LENGTH, allow_blank=False
    )


class _RedemptionAPIView(APIView):
    """Shared authority and HTTP shape for the redemption routes."""

    renderer_classes = (JSONRenderer,)
    authentication_classes = (SessionAuthentication,)
    permission_classes = (IsHouseholdParentOrChild,)

    def initial(self, request, *args, **kwargs):
        if request.method.lower() not in self.http_method_names:
            self.http_method_not_allowed(request)
        super().initial(request, *args, **kwargs)

    def get_permission_household(self, request):
        membership = resolve_active_membership(request.user, request)
        return None if membership is None else membership.household


class RedemptionListView(_RedemptionAPIView):
    """`GET /api/v1/redemptions/` and `POST /api/v1/redemptions/`."""

    http_method_names = ('get', 'head', 'post', 'options')

    def get_permissions(self):
        if self.request.method == 'POST':
            return [IsHouseholdChild()]
        return [IsHouseholdParentOrChild()]

    def get(self, request):
        membership = resolve_active_membership(request.user, request)
        household = None if membership is None else membership.household
        redemptions = Redemption.objects.filter(household=household)

        if membership is not None and membership.role == Membership.Role.CHILD:
            redemptions = redemptions.filter(child=request.user)
            serializer_class = RedemptionChildSerializer
        else:
            serializer_class = RedemptionParentSerializer

        return Response(serializer_class(redemptions, many=True).data)

    def post(self, request):
        membership = resolve_active_membership(request.user, request)
        if membership is None or membership.role != Membership.Role.CHILD:
            raise PermissionDenied(PERMISSION_DENIED_DETAIL)

        serializer = RedemptionWriteSerializer(data=request.data)
        serializer.is_valid(raise_exception=True)

        try:
            redemption, created = redeem_reward(
                household=membership.household,
                child=request.user,
                reward_id=serializer.validated_data['reward'],
                idempotency_key=serializer.validated_data['idempotency_key'],
            )
        except Reward.DoesNotExist:
            raise NotFound() from None
        except (Membership.DoesNotExist, RewardAuthorityError):
            raise PermissionDenied(PERMISSION_DENIED_DETAIL) from None
        except InsufficientPointsError:
            raise serializers.ValidationError(
                {'reward': [INSUFFICIENT_POINTS_DETAIL]}
            ) from None

        response_status = status.HTTP_201_CREATED if created else status.HTTP_200_OK
        return Response(
            RedemptionChildSerializer(redemption).data, status=response_status
        )


class _RedemptionActionView(_RedemptionAPIView):
    """One parent-only redemption transition, written exactly once."""

    http_method_names = ('post', 'options')
    permission_classes = (IsHouseholdParent,)
    service_fn = None

    def post(self, request, pk):
        membership = resolve_active_membership(request.user, request)
        if membership is None or membership.role != Membership.Role.PARENT:
            raise PermissionDenied(PERMISSION_DENIED_DETAIL)

        try:
            redemption = self.service_fn(
                household=membership.household,
                parent=request.user,
                redemption_id=pk,
            )
        except Redemption.DoesNotExist:
            # Covers missing, foreign-household, and already-decided rows
            # alike: a stale or repeated action is refused the same way a
            # row that never existed is, and neither answer confirms which.
            raise NotFound() from None
        except (Membership.DoesNotExist, RewardAuthorityError):
            raise PermissionDenied(PERMISSION_DENIED_DETAIL) from None

        return Response(RedemptionParentSerializer(redemption).data)


class RedemptionFulfilView(_RedemptionActionView):
    """`POST /api/v1/redemptions/<pk>/fulfil/`."""

    service_fn = staticmethod(fulfil_redemption)


class RedemptionCancelView(_RedemptionActionView):
    """`POST /api/v1/redemptions/<pk>/cancel/`."""

    service_fn = staticmethod(cancel_redemption)
